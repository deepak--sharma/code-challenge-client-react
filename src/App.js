import { useEffect, useState } from 'react';


const App = (props) => {
  const [items, setItems] = useState({});
  const { data } = props;
  useEffect(() => {
    const request = () =>
      data.forEach((product) => {
        console.log("product is ", product);
        fetch(`http://localhost:8081/temperature/${product.id}`)
          .then((response) => response.json())
          .then((response) => {
            console.log("items are ", items)
            console.log("new resp is ", response)
            console.log("product temperature is ", product.name);
            setItems((prevItems) => ({
              ...prevItems,
              [product.id]: {
                ...product,
                ...response,
              },
            }))
          }
          );
      });

    setInterval(request, 5000);

    request();
  }, []);

  return (
    <div className="App">
      <h2>Beers</h2>
      <div>{items & items[1]}</div>
      <table>
        <thead>
          <tr>
            <th align="left">Product</th>
            <th align="left">Temperature</th>
            <th align="left">Status</th>
          </tr>
        </thead>
        <tbody>
          {Object.keys(items).map((itemKey) => (
            <tr key={items[itemKey].id}>
              <td width={150}>{items[itemKey].name}</td>
              <td width={150}>{items[itemKey].temperature}</td>
              <td width={150}>
                {items[itemKey].temperature <
                  items[itemKey].minimumTemperature && <span>too low</span>}
                {items[itemKey].temperature >
                  items[itemKey].maximumTemperature && <span>too high</span>}
                {items[itemKey].temperature <=
                  items[itemKey].maximumTemperature &&
                  items[itemKey].temperature >=
                  items[itemKey].minimumTemperature && <span>all good</span>}
              </td>
            </tr>
          ))}
        </tbody>
      </table>
    </div>
  );
}

export default App;
