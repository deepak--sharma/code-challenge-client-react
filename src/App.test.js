import App from './App';
import renderer from 'react-test-renderer';
import { shallow } from 'enzyme';
import React, { useState as useStateMock } from 'react'

let container;
const messages = {
    good: "all good",
    low: "too low",
    high: "too high"
}
const beerItem = [{
    id: '1',
    name: 'Pilsner',
    minimumTemperature: 4,
    maximumTemperature: 6,
}, {
    id: '2',
    name: 'IPA',
    minimumTemperature: 5,
    maximumTemperature: 6,
},]
describe('<App />', () => {
    const setState = jest.fn()
    beforeEach(() => {
        jest.spyOn(React, 'useState')
            .mockImplementationOnce(() => useStateMock({ beerItem }));       
        const mockResponse = { id: "1", temperature: 4 };
        jest.spyOn(global, 'fetch').mockResolvedValue({
            json: jest.fn().mockResolvedValue(mockResponse)
        });
        container = document.createElement('div');
        document.body.appendChild(container);
    });
    afterEach(() => {
        // cleanup on exiting
        //unmountComponentAtNode(container);
        container.remove();
        container = null;
    });
    it('renders correctly', () => {
        const tree = renderer
            .create(<App data={beerItem} />)
            .toJSON();
        expect(tree).toMatchSnapshot();
    });
    it('should render the message', () => {
        const wrapper = shallow(<App data={beerItem} />);
        console.log("wrapper is ", wrapper.debug());
        expect(wrapper.find('tr').children()).toHaveLength(6);
        //I expected this to pass but somehow the wrapper is rendering empty table, see console log at line no 49
        //expect(wrapper.find('tr').children().contains(<td>Pilsner</td>)).toBeTruthy();
    });
});