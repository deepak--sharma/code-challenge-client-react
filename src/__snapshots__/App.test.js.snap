// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> renders correctly 1`] = `
<div
  className="App"
>
  <h2>
    Beers
  </h2>
  <div>
    0
  </div>
  <table>
    <thead>
      <tr>
        <th
          align="left"
        >
          Product
        </th>
        <th
          align="left"
        >
          Temperature
        </th>
        <th
          align="left"
        >
          Status
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          width={150}
        />
        <td
          width={150}
        />
        <td
          width={150}
        />
      </tr>
    </tbody>
  </table>
</div>
`;
